ssh/tailssh/incubator.go

```diff
@@ -304,6 +293,89 @@ func beIncubator(args []string) error {
 	return err
 }
 
+// TODO(andrew-d): verify that this works in more configurations before
+// enabling by default.
+const assertDropPrivileges = false
+
+// dropPrivileges contains all the logic for dropping privileges to a different
+// UID, GID, and set of supplementary groups. This function is
+// security-sensitive and ordering-dependent; please be very cautious if/when
+// refactoring.
+//
+// WARNING: if you change this function, you *MUST* run the TestDropPrivileges
+// test in this package as root on at least Linux, FreeBSD and Darwin. This can
+// be done by running:
+//
+//	go test -c ./ssh/tailssh/ && sudo ./tailssh.test -test.v -test.run TestDropPrivileges
+func dropPrivileges(logf logger.Logf, wantUid, wantGid int, supplementaryGroups []int) error {
+	fatalf := func(format string, args ...any) {
+		logf(format, args...)
+		os.Exit(1)
+	}
+
+	euid := os.Geteuid()
+	egid := os.Getegid()
+
+	if runtime.GOOS == "darwin" || runtime.GOOS == "freebsd" {
+		// On FreeBSD and Darwin, the first entry returned from the
+		// getgroups(2) syscall is the egid, and changing it with
+		// setgroups(2) changes the egid of the process. This is
+		// technically a violation of the POSIX standard; see the
+		// following article for more detail:
+		//    https://www.usenix.org/system/files/login/articles/325-tsafrir.pdf
+		//
+		// In this case, we add an entry at the beginning of the
+		// groupIDs list containing the expected gid if it's not
+		// already there, which modifies the egid and additional groups
```

Comment:
"If it's not already there" - on compliant platforms, is it permitted to repeat the egid in the supplemental groups list? If not, I would defensively make that a fatal error, just to reduce the universe of possible states we have to deal with (and thus the universe of possible bugs stemming from different OSes's interpretation/violation of posix)

Reply:
I tested both Darwin and FreeBSD, and it appears to be permitted to repeat the egid in the set of supplemental groups.  This behaviour also matches what Samba does [here](https://github.com/samba-team/samba/blob/master/source3/lib/system.c#L851-L863), which makes me a bit more confident that it's correct.

Reply:
Okay, fair enough.

Incidentally, that Samba code might point to why this behaves differently on Darwin: the comment suggests that darwin ignores the first supplemental gid entirely. So, it's possible on darwin prior to this change we were actually dropping one of the supplemental gids?

Reply:
Yeah, also that 😬 Yay for fixing bugs as well, I guess?